i6_models/primitives/specaugment.py

```diff
@@ -0,0 +1,118 @@
+import torch
+
+
+def _mask(tensor: torch.Tensor, batch_axis: int, axis: int, pos: torch.Tensor, max_len: int) -> torch.Tensor:
+    """
+    :param tensor: e.g. [B, ..., A, ...] but arbitrary axis order
+    :param batch_axis: index of the batch axis
+    :param axis: which axis A to mask
+    :param pos: at which positions along axis to start the mask (size [B])
+    :param max_len: mask length drawn uniformly from [0, max_len]
+    """
+    batch_dim = tensor.shape[batch_axis]
+    dim = tensor.shape[axis]
+    amount = torch.randint(low=1, high=max_len + 1, size=(batch_dim,), dtype=torch.int32).to(device=tensor.device)
+    pos2 = torch.min(pos + amount, torch.tensor([dim] * batch_dim).to(device=tensor.device))
+    idxs = torch.arange(0, dim).to(device=tensor.device).unsqueeze(0)  # [1,dim]
+    pos_bc = pos.unsqueeze(1)  # [B,1]
+    pos2_bc = pos2.unsqueeze(1)  # [B,1]
+    cond = torch.logical_and(torch.greater_equal(idxs, pos_bc), torch.less(idxs, pos2_bc))  # [B,dim]
+    if batch_axis > axis:
+        cond = cond.transpose(0, 1)  # [dim,B]
+    cond = torch.reshape(
+        cond, shape=[tensor.shape[i] if i in (batch_axis, axis) else 1 for i in range(len(tensor.shape))]
+    )
+    tensor = torch.where(cond, 0.0, tensor)
+    return tensor
+
+
+def _random_mask(tensor: torch.Tensor, batch_axis: int, axis: int, min_num: int, max_num: int, max_len: int):
+    """
+    Mask tensor along axis using N in [min_num, max_num] masks of length [0, max_len]
+
+    :param tensor: e.g. [B, ..., A, ...] but arbitrary axis order
+    :param batch_axis: index of the batch axis
+    :param axis: which axis to mask
+    :param min_num: minimum number of masks
+    :param max_num: maximum number of masks
+    :param max_amount: mask length drawn uniformly from [0, max_amount]
+    """
+
+    batch_dim = tensor.shape[batch_axis]
+    num_masks = torch.randint(min_num, max_num, size=(batch_dim,)).to(device=tensor.device)  # [B]
+
+    z = -torch.log(-torch.log(torch.rand((batch_dim, tensor.shape[axis])).to(device=tensor.device)))  # [B,dim]
+    _, indices = torch.topk(z, num_masks.max().item(), dim=1)
+
+    for i in range(num_masks.max().item()):
+        tensor = _mask(tensor, batch_axis, axis, indices[:, i], max_len)
+    return tensor
```

Comment:
There are several things about this code that I don't understand.

First of all, `num_masks` has shape [B] but only the maximum value is ever used. This makes it so that the sampling is no longer uniform between `min_num` and `max_num`; for larger B the expected value will get closer to `max_num`. I would just use shape [1] and remove the `.max()` calls. Also it doesn't seem to be necessary that this is a tensor, let alone on-device, since the value is always retrieved via `.item()`
```
    import random
    num_masks = random.randint(min_num, max_num)
```

Next, the construction of `z` and `indices` makes no sense to me. The two `-torch.log` transformations don't change the result indices of `top_k` in `z` so they can just be removed. And then, since `z` just consists of random numbers anyway, the indices are also random. So why not just set `indices` randomly directly?
```
    indices = [random.sample(range(tensor.shape[axis]), num_masks) for _ in range(batch_dim)]
    indices = torch.tensor(indices, device=tensor.device)
```
So overall I would change this to:
```suggestion
    import random

    num_masks = random.randint(min_num, max_num)

    indices = [random.sample(range(tensor.shape[axis]), num_masks) for _ in range(batch_dim)]  # [B, num_masks]
    indices = torch.tensor(indices, device=tensor.device)

    for i in range(num_masks):
        tensor = _mask(tensor, batch_axis, axis, indices[:, i], max_len)
    return tensor
```

Reply:
@mmz33 this is a question for you

Reply:
regarding sampling part, this was implemented by @albertz and he provided these links for reference:
* https://github.com/tensorflow/tensorflow/issues/9260                                                                                                                                                    
* https://timvieira.github.io/blog/post/2014/08/01/gumbel-max-trick-and-weighted-reservoir-sampling/   
```
This is also the TF code:                                                                                                   
z = -tf.log(-tf.log(tf.random_uniform((n_batch, tf.shape(x)[axis]), 0, 1)))                                                                                                                               
  _, indices = tf.nn.top_k(z, num if isinstance(num, int) else tf.reduce_max(num))  
```
So i understand correctly, this is a better sampling method but i am not sure about its influence here. I guess @albertz can comment on this.

Reply:
So the original code is this (as you find in the linked TF issue):
```python
z = -tf.log(-tf.log(tf.random_uniform(tf.shape(logits),0,1)))
_, indices = tf.nn.top_k(logits + z,K)
```

We also use the same code in the TF `ChoiceLayer` when you want to randomly sample from the distribution instead of taking the top_k.

I then removed the `logits` because I wanted it to be random sampling without replacement from a uniform distribution (logits = 0 everywhere).

I think I just kept the `-log(-log(..))` to keep it clear where this comes from and to potentially add some logits again in case you don't want to sample from uniform.

Now, without logits (or logits=0), sure you could remove the `-log(-log(..))`.